[OLD PR - DONT MERGE] Feature/add bml wfca request and response commands

morantr opened the pull request:
Remove unnecessary function that parses WFA-CA command and printed
to the console dummy return string, and drill down WFA-CA string command from bml to the controller, with double reply mechanism.
Also, add a new wfa_ca class to concentrate all WFA-CA activity.
#121

morantr committed:
common: btlvf: Add bml WFA-CA commands

Add WFA-CA command and response yaml to btlvf and its auto-generated
code.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: bml: Drill down WFA-CA command to the controller

1. Remove unnecessary function that parses WFA-CA command and printed
to the console dummy return string.

2. Drill down WFA-CA string command from bml to the controller.

3. Added command with response callback mechanism.
According to the Test Suit Control API specifications, UCC should get
two strings in return. The first on is "Running" which means that the
system received the command, and it started to do what has been requested,
and the second return is a "Complete" status that means that the system
succeded to perform what it was requested.
All of the replies which will be received to the will be routed to callback
function, which will print it.

4. Add new wfa_ca class to concentrate all WFA-CA activity.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
bml/framework: Change of naming

Change text "wfca" to "wfa_ca" in all the code.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

simonbaren commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1018,6 +805,8 @@ void cli_bml::events_update_to_socket_cb(const struct BML_EVENT *event)
     cli_bml::events_update_cb(event, false);
 }
 
+void cli_bml::wfa_ca_reply_to_console_cb(const char *buffer) { std::cout << buffer << std::endl; }
```
I see that the response is simple STDOUT. Why do you need all this complex callback and pointers mechanism? can't you just print the response once you get it?
It will simplify most of the logic you added to handle_response (pending_second, promises etc....)

morantr replied:
Because we need to handle two responses.
One response should be return at most in 1 second.
The second response may come after at most 120 seconds.  I can't block the cli between the two responses. So I must use callback which is similar to a response that is an "event" like.

tomereli replied:
Reopening since we decieded to go with a different approach of handling everything in the callback.

tomereli commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -527,26 +527,28 @@ int bml_get_slave_restricted_channels(BML_CTX ctx, uint8_t *restricted_channels,
 }
 
-int bml_wfca_controller(BML_CTX ctx, const char *cmd, char *ret_buf, int ret_buf_size)
+int bml_wfa_ca_controller(BML_CTX ctx, const char *command, int command_len,
```
It is generally better to do name changes in separate commit / PR as it adds more overhead to the reviewer, especially for big PRs. But take this as a note for next time, don't change it now.

morantr replied:
It is already in a different commit...

tomereli commented on controller/src/beerocks/master/wfa_ca.h:
```diff
@@ -0,0 +1,51 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2016-2019 Intel Corporation
```
2019

morantr replied:
Did you mean to change  "2016-2019" to "2019"?

tomereli replied:
Still not `2019` - please change

morantr replied:
I changed it... I don't understand how you still see 2016-2019

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,171 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2016-2019 Intel Corporation
```
2019

morantr replied:
Did you mean to change  "2016-2019" to "2019"?

arnout commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -527,15 +527,17 @@ int bml_get_slave_restricted_channels(BML_CTX ctx, uint8_t *restricted_channels,
 
-int bml_wfca_controller(BML_CTX ctx, const char *cmd, char *ret_buf, int ret_buf_size)
+int bml_wfca_controller(BML_CTX ctx, const char *command, int command_len, const char **first_reply,
+                        BML_WFA_CA_CB second_reply_cb)
```
In keeping with https://github.com/prplfoundation/prplMesh/issues/132#issuecomment-516061919 I think it would be better to remove the `ret_buf` argument entirely and instead always rely on event callbacks for the responses.

Also I think the callback should be a fixed member that gets registered once like the event callback, so not an argument to this function.

arnout commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -587,6 +587,12 @@ typedef void (*BML_STATS_UPDATE_CB)(const struct BML_STATS_ITER *);
+ * WFA_CA reply function. When registered, the function will be called after 
+ * the second reply of WFA-CA command. 
+ */
+typedef void (*BML_WFA_CA_CB)(const char *buffer);
```
I think we could have used an event instead, but adding a new callback type is OK as well.

tomereli replied:
I think you should add length argument - We never pass a pointer to a buffer without a length in C code, according to SDL.

morantr replied:
> I think we could have used an event instead, but adding a new callback type is OK as well.

BML events are actually the same. They use callback function as well. The use of them is for Beerocks notifications when something happened on the system and has one register function for all events. It is not fit here.

morantr replied:
> I think you should add length argument - We never pass a pointer to a buffer without a length in C code, according to SDL.

It is not needed. I pass to the callback const char* so it cannot be modified.
The size is also irrelevant because the string just being printed.
No security holes here. Guarantee!

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml:
```diff
@@ -345,3 +345,23 @@ cACTION_BML_TRIGGER_CHANNEL_SELECTION_REQUEST:
+    _type: uint16_t
+    _length_var: True
+  reply:
+    _type: uint8_t
```
Since this is going to be an actual string that is going to be printed, would it be possible to make this a char instead?

Same for the command, in fact.

If if it is a char instead of a uint8_t, it would be good to add an explicit comment saying whetther a terminating 0 is included or not (it isn't, I suppose).

tomereli replied:
As you said it would be better to have it as `char` instead of `uint8_t`. This way you get the `set_reply()` APIs which take care of the null termination (so yes).

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -83,6 +83,9 @@ bml_internal::~bml_internal() {}
 int bml_internal::send_bml_cmdu(int &result, uint8_t action_op)
 {
+    // initialize value
+    result = false;
```
This is an API-breaking change (before, result was only updated on success; now it is always updated). It should definitely be done in a separate commit with an explanation why.

In this case, it doesn't make much of a difference (especially since AFAIU there are no actual BML users that could be affected), so (nitpick).

morantr replied:
No, it is not a big deal change.
Just initializing the result integer to default value of "false", so if the function fails, it will return "false". If the function success it takes the value from the response.

arnout replied:
It doesn't return false; it overwrites the variable which is given as a reference. So this is really an API change.

But as I said, I'm sure it will not affect any existing users. Just for the future: be careful with API changes.

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -829,6 +832,48 @@ int bml_internal::process_cmdu_header(cmdu_vs_action_header_t beerocks_header,
+                return false;
+            }
+
+            char *reply_ptr = (char *)(&std::get<1>(reply_tuple));
```
I always feel uncomfortable when I see a C-style cast in C++ code...

tomereli replied:
There would be no need for this once it is changed in the yaml to `char`.
The code will be much simpler:
```
auto reply = response->reply(); //reply is a const std::string
```

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -829,6 +832,48 @@ int bml_internal::process_cmdu_header(cmdu_vs_action_header_t beerocks_header,
+
+                // Must use sleep here to prevent a race between the first reply to the second reply
+                // Without sleep here, the second reply may be printed first instead of the first
+                // reply, if both replies were sent at the same time.
```
Sleeping is never a solution for a race condition, it is wiping it under the carpet so it can show up at the worst possible time.

By always printing results in the callback, avoid this issue entirely because you never need to synchronise the command with the response.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,91 @@
+    case COMPLETE:
+        return "COMPLETE";
+    default:
+        break;
```
There should be some error logging here.

Also, (nitpick), I generally prefer to use the same pattern for all switch cases, so return here as well instead of break (and no return at end of function).

tomereli replied:
also, function name should fit what it does - something like `wfa_ca_status_to_string()`

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,91 @@
+{
+    std::transform(command.begin(), command.end(), command.begin(), ::toupper);
+
+    if (command == "CA_GET_VERSION") {
```
A big sequence of comparisons like this is nicer as a lookup table with callback functions. In this case, it could be just a string/enum mapping. However, this function is only used within handle_wfa_ca_message() to be able to use a switch there - so you could just as well do the table conditions in there and forget about the enum entirely.

So something like:

```
static const struct {
    const char *command;
    std::string (*handler) (std::string command, std::vector<std::string> params);
} = {
  {"CA_GET_VERSION", [](std::string command, std::vector<std::string> params) { return "VERSION 1"; } },
...
};
```

But maybe I'm overengineering and this is simpler.

tomereli replied:
For me this is overengineering and I would prefer to keep it as Moran did:)

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,91 @@
+    }
+
+    char *reply_ptr = (char *)(&std::get<1>(reply_tuple));
+    string_utils::copy_string(reply_ptr, reply_string.c_str(), reply_string.size() + 1);
```
Oh, so the 0 terminator *is* included then?

morantr replied:
it was

arnout reviewed (changes requested):
The major issue is that the race condition is not handled properly.

Some minor additional comments about the commit messages:

* The second commit contains 4 points; that is typically indicative that it would be better to split up into 4 separate commits.
* The third commit is a kind of cleanup of existing code. It is generally better to do such things as the first commit in the series, because (1) it could be merged separately before the rest is approved and (2) it makes it easier to understand the big commit (commit 2 in this case). However, it's sometimes difficult to reorder things like that, so OK to keep it as is.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2565,13 +2610,43 @@ int bml_internal::get_restricted_channels(uint8_t *restricted_channels, const st
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->alloc_command(command_len);
```
Same here - after you change to char in the yaml:
```
request->set_command(std::string(command), command_len);
```
No need for all of the fancy tuples and buffers, all taken care inside the TLVF.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2565,13 +2610,43 @@ int bml_internal::get_restricted_channels(uint8_t *restricted_channels, const st
 }
 
-int bml_internal::wfca_agent(const char *cmd, char *ret_buf, int ret_buf_size)
+int bml_internal::wfa_ca_agent(const char *cmd, char *ret_buf, int ret_buf_size)
```
Why isn't there cmd_len argument here? It is only OK not to include length of a const char *buffer if it is hard coded in the code, for example `const char *message ="send1905_cmdu,1234"` - AFAIK this is not the case, the string is received from the testbed through a TCP socket.

morantr replied:
I only changed this function name. Currently, it is not used.
On Agent certification tasks we will probably change/remove it...

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -241,6 +243,11 @@ class bml_internal : public beerocks::socket_thread {
     uint16_t id                                                  = 0;
+
+    // wfa-ca members
+    std::string m_wfa_ca_controller_first_reply;
```
As both me and @arnout suggested, please replace the mechanism to use callbacks only which will solve the race conditions and make the code simpler and easier to understand.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -214,219 +214,6 @@ static void bml_utils_dump_conn_map(
     }
 }
 
-static void print_bml_wfca_ret_val(const std::string &func_name, int ret_val,
```
Deleting unused code should be a separate commit. Note for next time (nitpick)

morantr replied:
It was used, and I replaced it...

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1781,27 +1570,29 @@ int cli_bml::get_slave_restricted_channels(const std::string &hostap_mac)
-    std::string ret_str;
-    if (!wfca_dummy_ret_val_gen(cmd, ret_str)) {
-        ret = BML_RET_INVALID_ARGS;
+    const char *first_reply = nullptr;
```
We are changing this mechanism anyway - but note for the future - it is better to define a buffer with fixed length on the stack then passing a nullptr to be filled in a different function, since it makes it very clear that the resource does not have to be freed.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,171 @@
+    return std::string();
+}
+
+wfa_ca::eWfaCaCommand wfa_ca::wfa_ca_command(std::string command)
```
argument should be const if possible

tomereli replied:
continueing the previous comment, the function name should be `wfa_ca_command_from_string()` like we have in other areas in our codebase.

morantr replied:
I have chosen to not pass it as const reference since I am changing it inside the function (transform).
Changed it to wfa_ca_command_from_string() though.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,171 @@
+        /* code */
+        break;
+    case eWfaCaCommand::DEV_CONFIGURE_IE:
+        /* code */
```
I prefer a comment starting with a `TODO` for each of these, with a short explanation on what should be done if it isn't obvious (nitpick)

morantr replied:
VSCODE add automatically /* code */
I changed it, but didn't add an explanation.
It is obvious that what needs to be done is to handle the command...

tomereli commented on controller/src/beerocks/master/wfa_ca.h:
```diff
@@ -0,0 +1,51 @@
+                          db &database, task_pool &tasks);
+
+private:
+    enum eWfaCaStatus : uint8_t { RUNNING, INVALID, ERROR, COMPLETE };
```
Better to use enum class here since there is no prefix for the enum values pointing on which enum it is.

morantr committed:
framework: tlvf: change char string setter prototype

When defining an array of chars on yaml file, the tlvf create setter to
a string that receives char*.
Change it to const char* since it is a constant.
Updated auto-generated files as well.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2565,9 +2590,28 @@ int bml_internal::get_restricted_channels(uint8_t *restricted_channels, const st
+                                    BML_WFA_CA_CB reply_cb)
 {
-    // TODO: implement function
+    //CMDU message
```
comment is redundant (nitpick)

morantr replied:
removed it

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2565,9 +2590,28 @@ int bml_internal::get_restricted_channels(uint8_t *restricted_channels, const st
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->set_command(command, command_len);
```
you could have just used `request->set_commnd(std::string(command, command_len))` and avoid the change in the TLVF (nitpick)

morantr replied:
The change in the TLVF is correct regardlessly. Besides using the pointer constructor saves an unnecessary construction of std::string.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -2565,9 +2590,28 @@ int bml_internal::get_restricted_channels(uint8_t *restricted_channels, const st
+
+    request->set_command(command, command_len);
+
+    m_cbWfaCaReply = reply_cb;
```
shouldn't we check that `reply_cb` is not `nullptr`?

morantr replied:
I don't know how I missed it

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1783,25 +1572,19 @@ int cli_bml::get_slave_restricted_channels(const std::string &hostap_mac)
-    }
-    print_bml_wfca_ret_val("wfca_agent", ret, ret_str);
-
+    // TODO
```
please elaborate on what needs to be done. If possible, add a task for it and refer to it here in the TODO comment

morantr replied:
I don't know currently what has to be done. This will be part of the agent certification.
It is irrelevant now, and I don't think it is a good idea to put effort into it now.
I didn't make changes here.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,33 @@
+void wfa_ca::handle_wfa_ca_message(
+    Socket *sd, std::shared_ptr<beerocks_message::cACTION_HEADER> beerocks_header,
+    ieee1905_1::CmduMessageRx &cmdu_rx, ieee1905_1::CmduMessageTx &cmdu_tx, db &database,
+    task_pool &tasks)
```
I don't see where the input parameters are used, for example `tasks` and `database`. Are these added for future reasons?

morantr replied:
yes

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,33 @@
+
+    if (!request) {
+        LOG(ERROR) << "addClass cACTION_BML_WFA_CA_CONTROLLER_REQUEST failed";
+        return;
```
Can we change the function to return `bool` so the error can also be handled in the future by the caller if necessary?

morantr replied:
We can, but really there is nothing to be handled.

tomereli replied:
Maybe in the future we will add handling.

arnout replied:
:-1: to that. I think it's a bad idea to add stuff for "maybe in the future we will".

For error handling, it only makes sense to add a return value that indicates error if the error is not handled yet. The error has been handled here (by logging an error); there is nothing more that can be done by the caller.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,33 @@
+        return;
+    }
+
+    auto command_tuple = request->command(0);
```
I guess the tuple is used here since the `command` is defined as `uint8_t` in the yaml file - if so, please change it so we can use the string APIs from the TLVF here.

arnout replied:
Currently tlvf doesn't have special handling for char*. It really is treated as an array of elements of type char, so you get the tuples and you need to do the ugly casting like below.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,33 @@
+#include "wfa_ca.h"
```
This file is missing our license header

morantr replied:
This is weird. In my code its exist.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -1,3 +1,11 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
```
Now I see the license snippet was added in the last commit, please squash to the commit which added this file

morantr replied:
About it... I don't think it is right to squash the two commits because of license comment.
They are two different logic units and should be in separate commits.

tomereli replied:
It was a neatpick anyway, so ok

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,89 @@
+
 using namespace beerocks;
 
+std::string wfa_ca::wfa_ca_status_to_string(eWfaCaStatus status)
```
return type should be `const std::string`

morantr replied:
Why? It is important to use const if you care that the caller will change it.
This is why we don't return "const bool" from other functions for example.

tomereli replied:
Because you return a const string so you add an unnecessary copy if the return value is not const.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,89 @@
+    return std::string();
+}
+
+wfa_ca::eWfaCaCommand wfa_ca::wfa_ca_command_from_string(std::string command)
```
argument type should be `const std::string`

morantr replied:
No, I use transform function to change it.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,89 @@
+    return WFA_CA_COMMAND_MAX;
+}
+
+void wfa_ca::reply_invalid_error(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx,
```
This function should return `bool` since it can fail on error conditions.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,89 @@
+
+    std::string reply_string =
+        "status," + wfa_ca_status_to_string(status) + ",errorCode," + description;
+    reply(sd, cmdu_tx, reply_string, false);
```
once you change it to `bool`, change the last line to `return reply(...)`

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,89 @@
+    }
+
+    if (!response->set_reply(reply_string)) {
+        LOG(ERROR) << "getting reply entry has failed!";
```
`setting reply ..`

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -8,8 +16,89 @@
+}
+
+bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::string &reply_string,
+                   bool pending_second_response)
```
Instead of calling it `pending_second_response`, call it `pending_additional_response` like you called it below.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -22,12 +111,54 @@ void wfa_ca::handle_wfa_ca_message(
-        return;
-    }
-    const char *command_ptr = (char *)(&std::get<1>(command_tuple));
+    const char *command_ptr = request->command();
```
instead of creating 2 variables, just use one - `std::string command = request->command();`

tomereli commented on controller/src/beerocks/master/wfa_ca.h:
```diff
@@ -21,6 +21,31 @@ class wfa_ca {
+    enum class eWfaCaStatus : uint8_t { RUNNING, INVALID, ERROR, COMPLETE };
+    static std::string wfa_ca_status_to_string(eWfaCaStatus status);
+
+    enum eWfaCaCommand : uint8_t {
```
consider changing to enum class (nitpick)

morantr replied:
I change it, although it doesn't really matter because the enums definitions are private anyway. and it can't be accessed from outside...

arnout replied:
Why do you specify the `uint8_t`? This is not a wire format definition, so by specifying `uint8_t` you're just increasing code size, no?

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -0,0 +1,33 @@
+        LOG(ERROR) << "getting command entry has failed!";
+        return;
+    }
+    const char *command_ptr = (char *)(&std::get<1>(command_tuple));
```
Please, no C-style casts.

morantr replied:
This piece of code has changed in the follow-up commit 998fe23e9e17a6a5c3653bfb522920d93526d8ed.
BTW c-like implicit casting notation is "( new_type ) expression".
CPP implict casting notation is "new_type ( expression )".
Since "char*(expression)" I put "char*" inside parantesis.

arnout replied:
`new_type (expression)` only works if `new_type` is a type name, not if it is a declarator. And it doesn't replace casts, it replaces conversions. `(int) expr` is a conversion to int, so it really is equivalent to `int(expr)`. For the `char*` case, however, even with the parenthesis, this is still parsed as a C-style cast.

More importantly, though, the programmer parses it as a C-style cast.

So, we should make this explicit (IMO) and call it a reinterpret_cast, because that's what it is.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -465,10 +465,10 @@ void cli_bml::setFunctionsMapAndArray()
         static_cast<pFunction>(&cli_bml::get_slave_restricted_channels_caller), 1, 1, STRING_ARG);
     insertCommandToMap(
-        "bml_wfca_controller", "<string_command>", "send wfca controller<string_command>",
-        static_cast<pFunction>(&cli_bml::bml_wfca_controller_caller), 1, 1, STRING_ARG);
```
This function was already renamed in the previous commit, so the previous commit is broken.

That's why I said that you should first do renames and only afterwards add new code :-).

@tomereli is it necessary to fix this?

tomereli replied:
we better fix it, yes. I'll help Moran.

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -22,12 +111,54 @@ void wfa_ca::handle_wfa_ca_message(
+    }
+
+    auto command_type     = wfa_ca_command_from_string(cmd_tokens_vec[0]);
+    auto command_type_str = cmd_tokens_vec[0];
```
It would be more logical to swap:

```
auto command_type_str = cmd_tokens_vec[0];
auto command_type = wfa_ca_command_from_string(command_type_str);
```

(nitpick, obviously)

morantr replied:
will change it

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -22,12 +111,54 @@ void wfa_ca::handle_wfa_ca_message(
+        break;
+    default:
+        auto err_description = "Invalid WFA-CA command type: '" + command_type_str + "'";
+        LOG(ERROR) << "Invalid WFA-CA command: " << command_type_str;
```
Reuse err_description here (nitpick).

morantr replied:
will fix

arnout reviewed (approved):
First of all, the commit messages are not OK. Is the documentation not clear enough? It should be one subject line, then an empty line, then paragraphs which are separated by empty lines. Just do `git log --oneline` to see the ugliness you have now. Please fix this wile rebasing.

Overall comment: I don't see the point of the pending_additional_response. IMO it just adds overhead. In addition, it still introduces a race condition: if a new command comes in while pending_additional_response is still true (i.e. while m_cbWfaCaReply is non-null), you'll get the wrong callback.

So, I repeat my earlier comment: I think the callback should be set once, globally, just like the event callback. And I think pending_additional_response and all related code should be removed.

Still, I don't think it's necessary to block the merge for this, so I'm approving after all. If you don't fix it now, please create an issue so we remember to fix it later.

So, please rebase (fixing the commit messages and maybe the small issues I still pointed out).

tomereli reviewed (approved):
Approved, but please fix Arnout's comments before merging.

tomereli commented:
New PR Opened - #138
